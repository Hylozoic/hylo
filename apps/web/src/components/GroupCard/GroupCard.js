import { Link } from 'react-router-dom'
import React from 'react'
import { useTranslation } from 'react-i18next'
import {
  DEFAULT_BANNER,
  DEFAULT_AVATAR,
  accessibilityDescription,
  accessibilityIcon,
  accessibilityString,
  visibilityDescription,
  visibilityIcon,
  visibilityString
} from 'store/models/Group'
import { bgImageStyle, cn } from 'util/index'
import { groupUrl, groupDetailUrl } from 'util/navigation'

import Icon from 'components/Icon'
import RoundImage from 'components/RoundImage'

import classes from './GroupCard.module.scss'

/*
  Each card needs
  - group title
  - group icon
  - group background (tiny, behind the group icon)
  - member sample (for avatars on right )
  - group geography descriptor (indigenous territory, location)

  TODO: Then is contents changed based on group type... perhaps passed in as a Content component
*/

export default function GroupCard ({ group, routeParams }) {
  const { t } = useTranslation()

  if (!group) return null

  return (
    <Link to={group.memberStatus === 'member' ? groupUrl(group.slug) : groupDetailUrl(group.slug, routeParams)}>
      <div className='flex relative rounded-lg p-4 bg-black shadow-xl hover:scale-102 transition-all duration-300'>
        <div className='flex gap-2 relative z-10 w-full justify-between'>
          <div className='flex flex-row gap-2'>
            <RoundImage url={group.avatarUrl || DEFAULT_AVATAR} size='50px' square className='shadow-xl' />
            <div className={cn('flex flex-row gap-2', classes.groupDetails)}>
              <div className='flex flex-col gap-0'>
                <span className={classes.groupName}>{group.name}</span>
                {group.memberCount ? <span className='text-xs text-white/80'>{group.memberCount} {t('Members')}</span> : ''}
              </div>
            </div>
          </div>
          <div className={cn(classes.groupStats)}>
            <div className={cn('flex flex-row gap-2 items-center h-full', classes.membershipStatus)}>
              <div className={classes.groupPrivacy}>
                <Icon name={visibilityIcon(group.visibility)} className={classes.privacyIcon} />
                <div className={classes.privacyTooltip}>
                  <div><strong>{t(visibilityString(group.visibility))}</strong> - {t(visibilityDescription(group.visibility))}</div>
                </div>
              </div>
              <div className={classes.groupPrivacy}>
                <Icon name={accessibilityIcon(group.accessibility)} className={classes.privacyIcon} />
                <div className={classes.privacyTooltip}>
                  <div><strong>{t(accessibilityString(group.accessibility))}</strong> - {t(accessibilityDescription(group.accessibility))}</div>
                </div>
              </div>
              {
                group.memberStatus === 'member'
                  ? <div className={classes.statusTag}><Icon name='Complete' className={classes.memberComplete} /> <b>{t('Member')}</b></div>
                  : group.memberStatus === 'requested'
                    ? <div className={classes.statusTag}><b>{t('Membership Requested')}</b></div>
                    : <div className='focus:text-foreground relative text-base border-2 border-foreground/20 hover:border-foreground/100 hover:text-foreground rounded-md p-2 bg-background text-foreground w-full transition-all scale-100 hover:scale-102 opacity-85 hover:opacity-100 flex items-center'><Icon name='CirclePlus' className={classes.joinGroup} /> <b>{t('Join')}</b></div>
              }
            </div>
          </div>
        </div>
        <div style={bgImageStyle(group.bannerUrl || DEFAULT_BANNER)} className='w-full h-full bg-cover bg-center rounded-lg absolute top-0 left-0 opacity-70' />
      </div>
    </Link>
  )
}
