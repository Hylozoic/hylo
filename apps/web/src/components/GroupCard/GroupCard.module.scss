.group-card {
  display: flex;
  border-bottom: 1px solid $color-rhino-10;
  padding-top: 6px;
  position: relative;
  padding-bottom: 6px;
  border-radius: 5px;
  color: white;
  padding: 15px;
  filter: drop-shadow(0px 5px 15px rgba(41, 64, 90, .2));
  margin-bottom: 15px;
  min-width: 0;
  transition: all .1s ease;
  top: 0px;

  &:hover {
    top: -2px;
    filter: drop-shadow(0px 5px 15px rgba(41, 64, 90, .5));
  }

  .group-description {
    display: block;
    width: 100%;
    text-decoration: none;
    color: white;
    max-height: 36px;
    overflow: hidden;

    span {
      overflow: hidden;
      text-overflow: ellipsis;
      display: block;
      text-decoration: none;
      color: white;
      font-size: 14px;
      line-height: 18px;
    }
  }

}

.group-image {
  margin-right: 15px;
  min-width: 50px;
  box-shadow: 0px 4px 15px rgba(41, 64, 90, .2);
  margin-left: -25px;
}

.group-details {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  justify-content: center;
  min-width: 0;
}

.group-name {
  color: white;
  @include font-bold;
  font-size: 20px;
  line-height: 20px;
  text-decoration: none;
  padding-right: 140px;
}

.group-stats {
  text-decoration: none;
  color: white;
  @include font-regular;
  font-size: 12px;
  opacity: .9;
  transition: opacity .25s ease;

  &:hover {
    opacity: 1;
  }

  .member-count {
    margin-right: 6px;
  }

  .membership-status {

    .group-privacy {
      margin-right: 8px;

      .privacy-icon {
        color: $color-white;
        font-size: 18px;
        position: relative;
        top: 3px;
      }

      .privacy-tooltip {
        display: none;
        position: absolute;
        width: 240px;
        top: 30px;
        right: 0;
        background-color: $color-black;
        font-size: 12px;
        padding: 5px 8px;
        border-radius: 5px;
        z-index: 1000;
      }

      &:hover {
        .privacy-tooltip {
          display: block;
        }
      }
    }

    .member-complete {
      color: rgba(255,255,255,1);
      margin-left: 1px;
    }

    .join-group {
      font-size: 18px;
      color: rgba(255,255,255,1);
      margin-left: 1px;
    }

    .status-tag {
      border: 1px solid rgba(255,255,255,.3);
      background-color: rgba(35, 64, 91, .2);
      height: 26px;
      line-height: 20px;
      border-radius: 20px;
      padding: 0 8px 0 2px;
      display: flex;
      align-items: center;
      color: rgba(255,255,255,1);
      font-size: 12px;

      b {
        margin-left: 5px;
        color: rgba(255,255,255,1);
      }
    }
  }
}

.group-link {
  text-decoration: none;

  &:hover {
    text-decoration: none;
  }
}

.card-wrapper {
  z-index: 1;
  display: flex;
  min-width: 0;
  align-items: center;
}
