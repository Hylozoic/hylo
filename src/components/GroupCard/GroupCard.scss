.card {
  border: 1px solid $color-cape-cod-10;
  border-radius: 4px;
  cursor: pointer;
  position: relative;
  color: black;
  right: 0px;
  height: 223px;
  transition: 0.25s;
  display: flex;
  flex-direction: column;
  &:hover {
    box-shadow: 0px 0px 6px 0px $color-black-10;
    text-decoration: none;
  }
}

.group-description {
  @include font-regular;
  padding: 0 15px;
  width: 472px;
  font-size: 12px;
  line-height: 1.5;
  color: $color-regent;
  opacity: 0.8;
  height: 4.5em;
  margin-bottom: 12px;
  display: -webkit-box;
  -webkit-line-clamp: 3;
  -webkit-box-orient: vertical;
  overflow: hidden;
  text-overflow: ellipsis;
}

.group-tags {
  height: 78px;
  width: 472px;
  overflow: hidden;
}

.tag-pill{
  font-family: Circular Std;
  text-transform: none;
  font-style: normal;
  font-weight: 400;
  font-size: 14px;
  line-height: 18px;
}

.constrained.card {
  border-top: 5px solid $color-cape-cod-10;
  padding-top: 10px;
}

.expanded,
.expanded:hover {
  box-shadow: 0px 4px 6px 0px $color-black-10;
}

.group-link {
  text-decoration: none;

  &:hover {
    text-decoration: none;
  }
}