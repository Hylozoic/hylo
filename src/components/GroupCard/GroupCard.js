import React from 'react'
import cx from 'classnames'
import './GroupCard.scss'
import GroupHeader from './GroupHeader'
import { Link } from 'react-router-dom'
import { groupUrl, groupDetailUrl } from 'util/navigation'
import Pill from 'components/Pill'
import { capitalize } from 'lodash'

/*
  Each card needs
  - group title
  - group icon
  - group background (tiny, behind the group icon)
  - member sample (for avatars on right )
  - group geography descriptor (indigenous territory, location)

  TODO: Then is contents changed based on group type... perhaps passed in as a Content component
  Default for normie groups is Topics, and if no topics, pass in description

*/

export default function GroupCard ({ memberships, group = {}, routeParams = {}, highlightProps = {}, className, expanded = false, constrained = false, onClick = () => {} }) {
  const topics = group.groupTopics.toModelArray()
  return <Link to={memberships.includes(group.id) ? groupUrl(group.slug) : groupDetailUrl(group.slug, routeParams)} styleName='group-link'>
    <div
      onClick={onClick}
      styleName={cx('card', { expanded }, { constrained })}
      className={className}>
      <GroupHeader
        {...group}
        group={group}
        routeParams={routeParams}
        highlightProps={highlightProps}
        constrained={constrained} />
      <div styleName='group-description'>
        {group.description}
      </div>
      <div styleName='group-tags'>
        {topics.map(topic => <Pill styleName='tag-pill' darkText label={capitalize(topic.topic.name.toLowerCase())} id={topic.id} />)}
      </div>
    </div>
  </Link>
}
