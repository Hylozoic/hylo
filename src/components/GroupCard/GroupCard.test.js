import GroupCard from './GroupCard'
import { shallow } from 'enzyme'
import React from 'react'
// memberships, group = {}, routeParams = {}, highlightProps = {}, className, expanded = false, constrained = false, onClick = () => {}
const props = {
  group: {
    id: 1,
    name: 'A Great Cause',
    slug: 'great-cause',
    groupTopics: {
      toModelArray: () => [
        { topic: {
          name: 'Life',
          id: '1wooooop'
        } },
        { topic: {
          name: 'Love',
          id: '2wooooop'
        } },
        { topic: {
          name: 'Light',
          id: '3wooooop'
        } },
        { topic: {
          name: 'LOLS',
          id: '4wooooop'
        } }
      ]
    },
    description: 'the description, which is long enough to add a "Read More" button, ' +
      'the description, which is long enough to add a "Read More" button, ' +
      'the description, which is long enough to add a "Read More" button, ' +
      'the description, which is long enough to add a "Read More" button, ' +
      'the description, which is long enough to add a "Read More" button, '
  },
  expanded: false,
  memberships: [12, 24, 25, 346]
}

it('matches last snapshot', () => {
  const wrapper = shallow(<GroupCard {...props} />)
  expect(wrapper).toMatchSnapshot()
})

it('matches last snapshot with different config', () => {
  const differentProps = {
    ...props,
    constrained: true
  }
  const wrapper = shallow(<GroupCard {...differentProps} />)
  expect(wrapper).toMatchSnapshot()
})
