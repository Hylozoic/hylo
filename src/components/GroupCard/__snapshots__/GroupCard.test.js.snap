// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches last snapshot 1`] = `
<Link
  data-stylename="group-link"
  to="/all/group/great-cause"
>
  <div
    data-stylename="card"
    onClick={[Function]}
  >
    <GroupHeader
      constrained={false}
      description="the description, which is long enough to add a \\"Read More\\" button, the description, which is long enough to add a \\"Read More\\" button, the description, which is long enough to add a \\"Read More\\" button, the description, which is long enough to add a \\"Read More\\" button, the description, which is long enough to add a \\"Read More\\" button, "
      group={
        Object {
          "description": "the description, which is long enough to add a \\"Read More\\" button, the description, which is long enough to add a \\"Read More\\" button, the description, which is long enough to add a \\"Read More\\" button, the description, which is long enough to add a \\"Read More\\" button, the description, which is long enough to add a \\"Read More\\" button, ",
          "groupTopics": Object {
            "toModelArray": [Function],
          },
          "id": 1,
          "name": "A Great Cause",
          "slug": "great-cause",
        }
      }
      groupTopics={
        Object {
          "toModelArray": [Function],
        }
      }
      highlightProps={Object {}}
      id={1}
      name="A Great Cause"
      routeParams={Object {}}
      slug="great-cause"
    />
    <div
      data-stylename="group-description"
    >
      the description, which is long enough to add a "Read More" button, the description, which is long enough to add a "Read More" button, the description, which is long enough to add a "Read More" button, the description, which is long enough to add a "Read More" button, the description, which is long enough to add a "Read More" button, 
    </div>
    <div
      data-stylename="group-tags"
    >
      <Pill
        darkText={true}
        data-stylename="tag-pill"
        label="Life"
      />
      <Pill
        darkText={true}
        data-stylename="tag-pill"
        label="Love"
      />
      <Pill
        darkText={true}
        data-stylename="tag-pill"
        label="Light"
      />
      <Pill
        darkText={true}
        data-stylename="tag-pill"
        label="Lols"
      />
    </div>
  </div>
</Link>
`;

exports[`matches last snapshot with different config 1`] = `
<Link
  data-stylename="group-link"
  to="/all/group/great-cause"
>
  <div
    data-stylename="card constrained"
    onClick={[Function]}
  >
    <GroupHeader
      constrained={true}
      description="the description, which is long enough to add a \\"Read More\\" button, the description, which is long enough to add a \\"Read More\\" button, the description, which is long enough to add a \\"Read More\\" button, the description, which is long enough to add a \\"Read More\\" button, the description, which is long enough to add a \\"Read More\\" button, "
      group={
        Object {
          "description": "the description, which is long enough to add a \\"Read More\\" button, the description, which is long enough to add a \\"Read More\\" button, the description, which is long enough to add a \\"Read More\\" button, the description, which is long enough to add a \\"Read More\\" button, the description, which is long enough to add a \\"Read More\\" button, ",
          "groupTopics": Object {
            "toModelArray": [Function],
          },
          "id": 1,
          "name": "A Great Cause",
          "slug": "great-cause",
        }
      }
      groupTopics={
        Object {
          "toModelArray": [Function],
        }
      }
      highlightProps={Object {}}
      id={1}
      name="A Great Cause"
      routeParams={Object {}}
      slug="great-cause"
    />
    <div
      data-stylename="group-description"
    >
      the description, which is long enough to add a "Read More" button, the description, which is long enough to add a "Read More" button, the description, which is long enough to add a "Read More" button, the description, which is long enough to add a "Read More" button, the description, which is long enough to add a "Read More" button, 
    </div>
    <div
      data-stylename="group-tags"
    >
      <Pill
        darkText={true}
        data-stylename="tag-pill"
        label="Life"
      />
      <Pill
        darkText={true}
        data-stylename="tag-pill"
        label="Love"
      />
      <Pill
        darkText={true}
        data-stylename="tag-pill"
        label="Light"
      />
      <Pill
        darkText={true}
        data-stylename="tag-pill"
        label="Lols"
      />
    </div>
  </div>
</Link>
`;
